rampwf/score_types/detection.py

```diff
@@ -239,55 +214,99 @@ def scp_single(y_true, y_pred, shape, minipatch=None):
     return score, n_true, n_pred
 
 
-def ospa_single(y_true, y_pred, cut_off=1, minipatch=None):
+def ospa(y_true, y_pred, minipatch=None):
+    """
+    Optimal Subpattern Assignment (OSPA) metric for IoU score.
+
+    This metric provides a coherent way to compute the miss-distance
+    between the detection and alignment of objects. Among all
+    combinations of true/predicted pairs, if finds the best alignment
+    to minimise the distance, and still takes into account missing
+    or in-excess predicted values through a cardinality score.
+
+    The lower the value the smaller the distance.
+
+    Parameters
+    ----------
+    y_true, y_pred : list of list of tuples
+    minipatch : list of int, optional
+        Bounds of the internal scoring patch (default is None)
+
+    Returns
+    -------
+    float: distance between input arrays
+
+    References
+    ----------
+    http://www.dominic.schuhmacher.name/papers/ospa.pdf
+
+    """
+    ospas = np.array([ospa_single(t, p, minipatch)
+                      for t, p in zip(y_true, y_pred)])
+
+    ious, n_count, n_total = ospas.sum(axis=0)
+
+    if n_total == 0:
+        return 0
+
+    return ious * n_count / n_total
+
+
+def ospa_single(y_true, y_pred, minipatch=None):
     """
     OSPA score on single patch. See docstring of `ospa` for more info.
 
     Parameters
     ----------
     y_true, y_pred : ndarray of shape (3, x)
         arrays of (x, y, radius)
-    cut_off : float, optional (default is 1)
-        penalizing value for wrong cardinality
     minipatch : list of int, optional
         Bounds of the internal scoring patch (default is None)
 
     Returns
     -------
-    float: distance between input arrays
+    (iou_sum, n_pred, n_total)
+        float - sum of ious of matched entries
+        int - sum of matched entries
+        int - sum of all entries
 
     """
     n_true = len(y_true)
     n_pred = len(y_pred)
-
-    n_max = max(n_true, n_pred)
-    n_min = min(n_true, n_pred)
+    n_total = n_true + n_pred
 
     # No craters and none found
     if n_true == 0 and n_pred == 0:
-        return 0
+        return 2, 2, 2
```

Comment:
Is this correct? As that would give an OSPA of 2\*2 / 2 = 2, and thus score of 1 - 2 = -1 for a single patch? 
(didn't think this through yet, just noticed the consequence of those values)

Reply:
but maybe a score on a single patch for this (no craters, non found) will maybe never make much sense.

The question is how to weight the fact this was done correctly (detecting there are no craters on this particular patch). When taking the aggregate of all patches, maybe we shouldn't weight this and just return 0,0,0 ?

Reply:
Yes, I would return 0, 0, 0